File: backend/src/services/status-effect.service.ts
import { GameState } from '../models/game-state';
import { StatusEffect, StatusTriggerMoment, EnemyTurnAction } from '../models/battle-state';
import { CardEffectService, EffectTarget } from './card-effect.service';
import { CardEffectRepository } from '../repo/card-effect-repo';
import { StatusRepository, StatusEffectDefinition } from '../repo/status-repo';
import { Companion } from '../models/companion';
import { BattleEnemy } from '../models/battle-state';

/**
 * Manages the lifecycle of status effects on companions and enemies:
 *   - applyStatus  : add or stack a status on a target
 *   - tickStatuses : fire all statuses matching a trigger moment and return
 *                    synthetic EnemyTurnAction entries for the popup
 *   - removeStatus : explicit removal
 */
export class StatusEffectService {
  private effectService = new CardEffectService();
  private effectRepo    = new CardEffectRepository();
  private statusRepo    = new StatusRepository();

  // ---------------------------------------------------------------------------
  // Apply
  // ---------------------------------------------------------------------------

  /**
   * Apply `stacks` of a status identified by `statusId` to every target.
   * If the target already carries the same status, stacks are added.
   */
  async applyStatus(
    targets: EffectTarget[],
    statusId: string,
    stacks: number,
    state: GameState
  ): Promise<GameState> {
    const def = await this.statusRepo.getById(statusId);
    if (!def) return state;

    const targetIds = new Set(targets.map(t => t.id));

    // Update companions
    const updatedCompanions = state.companions.map(c => {
      if (!targetIds.has(c.id)) return c;
      return this.upsertStatus(c, def, stacks) as Companion;
    });

    // Update enemies
    const updatedEnemies = (state.battle?.enemies ?? []).map(e => {
      if (!targetIds.has(e.id)) return e;
      return this.upsertStatus(e, def, stacks) as BattleEnemy;
    });

    return {
      ...state,
      companions: updatedCompanions,
      battle: state.battle
        ? { ...state.battle, enemies: updatedEnemies }
        : state.battle,
    };
  }

  // ---------------------------------------------------------------------------
  // Tick
  // ---------------------------------------------------------------------------

  /**
   * Fire every status whose `triggerMoment` matches `moment` on all
   * companions and enemies. Returns the updated state plus synthetic
   * EnemyTurnAction entries describing status tick damage (for the popup).
   */
  async tickStatuses(
    moment: StatusTriggerMoment,
    state: GameState
  ): Promise<{ state: GameState; actions: EnemyTurnAction[] }> {
    const actions: EnemyTurnAction[] = [];
    let workingState = state;

    // --- Tick companions ---
    for (const companion of workingState.companions) {
      if (companion.life <= 0) continue;
      const statuses = companion.statusEffects ?? [];
      const matching = statuses.filter(s => s.triggerMoment === moment);
      if (matching.length === 0) continue;

      for (const status of matching) {
        const def = await this.statusRepo.getById(status.id);
        if (!def) continue;

        const tickEffect = await this.effectRepo.getById(def.tickEffectId);
        if (!tickEffect) continue;

        // Override value with current stacks (e.g. poison damage = stacks).
        const resolvedEffect = { ...tickEffect, value: status.stacks };

        // Capture HP before
        const hpBefore = this.findCompanion(workingState, companion.id)?.life ?? 0;

        workingState = this.effectService.apply(
          resolvedEffect,
          companion,   // source (the status "acts" on the target)
          [this.findCompanion(workingState, companion.id)!],
          workingState
        );

        const hpAfter    = this.findCompanion(workingState, companion.id)?.life ?? 0;
        const dmg        = hpBefore - hpAfter;
        const killed     = hpAfter <= 0;

        if (dmg > 0 || killed) {
          actions.push({
            enemyId:     `status-${status.id}`,
            enemyName:   status.name,
            attackName:  `${status.name} (${status.stacks} stacks)`,
            targetId:    companion.id,
            targetName:  companion.name,
            damageDealt: dmg,
            killedTarget: killed,
          });
        }

        // Decay stacks
        workingState = this.decayStatus(workingState, companion.id, status.id, def.decayPerTick, 'companion');
      }
    }

    // --- Tick enemies ---
    for (const enemy of (workingState.battle?.enemies ?? [])) {
      if (enemy.life <= 0) continue;
      const statuses = enemy.statusEffects ?? [];
      const matching = statuses.filter(s => s.triggerMoment === moment);
      if (matching.length === 0) continue;

      for (const status of matching) {
        const def = await this.statusRepo.getById(status.id);
        if (!def) continue;

        const tickEffect = await this.effectRepo.getById(def.tickEffectId);
        if (!tickEffect) continue;

        // For enemy ticks, override target to 'wildMonster' so damage hits
        // the enemy instead of a companion.
        const resolvedEffect = { ...tickEffect, value: status.stacks, target: 'wildMonster' as const };

        const hpBefore = this.findEnemy(workingState, enemy.id)?.life ?? 0;

        workingState = this.effectService.apply(
          resolvedEffect,
          enemy,
          [this.findEnemy(workingState, enemy.id)!],
          workingState
        );

        const hpAfter = this.findEnemy(workingState, enemy.id)?.life ?? 0;
        const dmg     = hpBefore - hpAfter;

        // Log enemy status ticks in the battle log.
        if (dmg > 0) {
          workingState = {
            ...workingState,
            battle: {
              ...workingState.battle!,
              log: [
                ...workingState.battle!.log,
                `${status.name} deals ${dmg} damage to ${enemy.name} (${status.stacks} stacks).`,
              ],
            },
          };
        }

        // Decay stacks
        workingState = this.decayStatus(workingState, enemy.id, status.id, def.decayPerTick, 'enemy');
      }
    }

    return { state: workingState, actions };
  }

  // ---------------------------------------------------------------------------
  // Remove
  // ---------------------------------------------------------------------------

  removeStatus(targetId: string, statusId: string, kind: 'companion' | 'enemy', state: GameState): GameState {
    if (kind === 'companion') {
      return {
        ...state,
        companions: state.companions.map(c =>
          c.id !== targetId
            ? c
            : { ...c, statusEffects: (c.statusEffects ?? []).filter(s => s.id !== statusId) }
        ),
      };
    }
    return {
      ...state,
      battle: state.battle
        ? {
            ...state.battle,
            enemies: state.battle.enemies.map(e =>
              e.id !== targetId
                ? e
                : { ...e, statusEffects: (e.statusEffects ?? []).filter(s => s.id !== statusId) }
            ),
          }
        : state.battle,
    };
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private upsertStatus(
    target: Companion | BattleEnemy,
    def: StatusEffectDefinition,
    stacks: number
  ): Companion | BattleEnemy {
    const existing = (target.statusEffects ?? []).slice();
    const idx = existing.findIndex(s => s.id === def.id);

    if (idx >= 0) {
      // Stack onto existing
      existing[idx] = { ...existing[idx], stacks: existing[idx].stacks + stacks };
    } else {
      // New status instance
      const status: StatusEffect = {
        id:             def.id,
        name:           def.name,
        icon:           def.icon,
        stacks,
        turnsRemaining: null,
        triggerMoment:  def.triggerMoment,
        effectId:       def.tickEffectId,
      };
      existing.push(status);
    }

    return { ...target, statusEffects: existing };
  }

  private decayStatus(
    state: GameState,
    targetId: string,
    statusId: string,
    decay: number,
    kind: 'companion' | 'enemy'
  ): GameState {
    const updater = (statuses: StatusEffect[]): StatusEffect[] => {
      return statuses
        .map(s => {
          if (s.id !== statusId) return s;
          const newStacks = s.stacks - decay;
          const newTurns  = s.turnsRemaining != null ? s.turnsRemaining - 1 : null;
          return { ...s, stacks: newStacks, turnsRemaining: newTurns };
        })
        .filter(s => s.stacks > 0 && (s.turnsRemaining === null || s.turnsRemaining > 0));
    };

    if (kind === 'companion') {
      return {
        ...state,
        companions: state.companions.map(c =>
          c.id !== targetId
            ? c
            : { ...c, statusEffects: updater(c.statusEffects ?? []) }
        ),
      };
    }

    return {
      ...state,
      battle: state.battle
        ? {
            ...state.battle,
            enemies: state.battle.enemies.map(e =>
              e.id !== targetId
                ? e
                : { ...e, statusEffects: updater(e.statusEffects ?? []) }
            ),
          }
        : state.battle,
    };
  }

  private findCompanion(state: GameState, id: string): Companion | undefined {
    return state.companions.find(c => c.id === id);
  }

  private findEnemy(state: GameState, id: string): BattleEnemy | undefined {
    return state.battle?.enemies.find(e => e.id === id);
  }
}

File: backend-csharp/TheGameIsOn.API/Services/StatusEffectService.cs
using TheGameIsOn.API.Models;
using TheGameIsOn.API.Repositories;

namespace TheGameIsOn.API.Services;

public interface IStatusEffectService
{
    Task<GameState> ApplyStatus(List<IEffectEntity> targets, string statusId, int stacks, GameState state);
    Task<(GameState State, List<EnemyTurnAction> Actions)> TickStatuses(StatusTriggerMoment moment, GameState state);
    GameState RemoveStatus(string targetId, string statusId, string kind, GameState state);
}

/// <summary>
/// Manages the lifecycle of status effects on companions and enemies.
/// </summary>
public class StatusEffectService : IStatusEffectService
{
    private readonly ICardEffectService _effectService;
    private readonly ICardEffectRepository _effectRepo;
    private readonly IStatusRepository _statusRepo;

    public StatusEffectService(
        ICardEffectService effectService,
        ICardEffectRepository effectRepo,
        IStatusRepository statusRepo)
    {
        _effectService = effectService;
        _effectRepo = effectRepo;
        _statusRepo = statusRepo;
    }

    // -------------------------------------------------------------------------
    // Apply
    // -------------------------------------------------------------------------

    public async Task<GameState> ApplyStatus(
        List<IEffectEntity> targets, string statusId, int stacks, GameState state)
    {
        var def = await _statusRepo.GetByIdAsync(statusId);
        if (def is null) return state;

        var targetIds = new HashSet<string>(targets.Select(t => t.Id));

        var updatedCompanions = state.Companions.Select(c =>
            !targetIds.Contains(c.Id) ? c : UpsertStatus(c, def, stacks)
        ).ToList();

        var updatedEnemies = (state.Battle?.Enemies ?? new()).Select(e =>
            !targetIds.Contains(e.Id) ? e : UpsertStatusEnemy(e, def, stacks)
        ).ToList();

        return new GameState
        {
            Player = state.Player,
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = updatedCompanions,
            History = state.History,
            Battle = state.Battle is not null
                ? CloneBattle(state.Battle, b => b.Enemies = updatedEnemies)
                : null,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    // -------------------------------------------------------------------------
    // Tick
    // -------------------------------------------------------------------------

    public async Task<(GameState State, List<EnemyTurnAction> Actions)> TickStatuses(
        StatusTriggerMoment moment, GameState state)
    {
        var actions = new List<EnemyTurnAction>();
        var workingState = state;

        // --- Tick companions ---
        foreach (var companion in workingState.Companions.ToList())
        {
            if (companion.Life <= 0) continue;
            var statuses = companion.StatusEffects ?? new();
            var matching = statuses.Where(s => s.TriggerMoment == moment).ToList();
            if (matching.Count == 0) continue;

            foreach (var status in matching)
            {
                var def = await _statusRepo.GetByIdAsync(status.Id);
                if (def is null) continue;

                var tickEffect = await _effectRepo.GetByIdAsync(def.TickEffectId);
                if (tickEffect is null) continue;

                var resolvedEffect = tickEffect with { Value = status.Stacks };

                var currentComp = FindCompanion(workingState, companion.Id);
                if (currentComp is null) continue;
                var hpBefore = currentComp.Life;

                workingState = _effectService.Apply(
                    resolvedEffect, currentComp, new List<IEffectEntity> { currentComp }, workingState);

                var hpAfter = FindCompanion(workingState, companion.Id)?.Life ?? 0;
                var dmg = hpBefore - hpAfter;
                var killed = hpAfter <= 0;

                if (dmg > 0 || killed)
                {
                    actions.Add(new EnemyTurnAction
                    {
                        EnemyId = $"status-{status.Id}",
                        EnemyName = status.Name,
                        AttackName = $"{status.Name} ({status.Stacks} stacks)",
                        TargetId = companion.Id,
                        TargetName = companion.Name,
                        DamageDealt = dmg,
                        KilledTarget = killed
                    });
                }

                workingState = DecayStatus(workingState, companion.Id, status.Id, def.DecayPerTick, "companion");
            }
        }

        // --- Tick enemies ---
        foreach (var enemy in (workingState.Battle?.Enemies ?? new()).ToList())
        {
            if (enemy.Life <= 0) continue;
            var statuses = enemy.StatusEffects ?? new();
            var matching = statuses.Where(s => s.TriggerMoment == moment).ToList();
            if (matching.Count == 0) continue;

            foreach (var status in matching)
            {
                var def = await _statusRepo.GetByIdAsync(status.Id);
                if (def is null) continue;

                var tickEffect = await _effectRepo.GetByIdAsync(def.TickEffectId);
                if (tickEffect is null) continue;

                var resolvedEffect = tickEffect with { Value = status.Stacks, Target = CardEffectTarget.WildMonster };

                var currentEnemy = FindEnemy(workingState, enemy.Id);
                if (currentEnemy is null) continue;
                var hpBefore = currentEnemy.Life;

                workingState = _effectService.Apply(
                    resolvedEffect, currentEnemy, new List<IEffectEntity> { currentEnemy }, workingState);

                var hpAfter = FindEnemy(workingState, enemy.Id)?.Life ?? 0;
                var dmg = hpBefore - hpAfter;

                if (dmg > 0)
                {
                    workingState = new GameState
                    {
                        Player = workingState.Player,
                        Graph = workingState.Graph,
                        Cards = workingState.Cards,
                        Companions = workingState.Companions,
                        History = workingState.History,
                        Battle = CloneBattle(workingState.Battle!, b =>
                            b.Log = new List<string>(b.Log)
                            {
                                $"{status.Name} deals {dmg} damage to {enemy.Name} ({status.Stacks} stacks)."
                            }),
                        PendingAbilityChoices = workingState.PendingAbilityChoices
                    };
                }

                workingState = DecayStatus(workingState, enemy.Id, status.Id, def.DecayPerTick, "enemy");
            }
        }

        return (workingState, actions);
    }

    // -------------------------------------------------------------------------
    // Remove
    // -------------------------------------------------------------------------

    public GameState RemoveStatus(string targetId, string statusId, string kind, GameState state)
    {
        if (kind == "companion")
        {
            return new GameState
            {
                Player = state.Player,
                Graph = state.Graph,
                Cards = state.Cards,
                Companions = state.Companions.Select(c =>
                    c.Id != targetId ? c : CardEffectService.CloneCompanion(c, comp =>
                        comp.StatusEffects = (comp.StatusEffects ?? new()).Where(s => s.Id != statusId).ToList())
                ).ToList(),
                History = state.History,
                Battle = state.Battle,
                PendingAbilityChoices = state.PendingAbilityChoices
            };
        }

        return new GameState
        {
            Player = state.Player,
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = state.Companions,
            History = state.History,
            Battle = state.Battle is not null
                ? CloneBattle(state.Battle, b =>
                    b.Enemies = b.Enemies.Select(e =>
                        e.Id != targetId ? e : CloneBattleEnemy(e, be =>
                            be.StatusEffects = (be.StatusEffects ?? new()).Where(s => s.Id != statusId).ToList())
                    ).ToList())
                : null,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private static Companion UpsertStatus(Companion target, StatusEffectDefinition def, int stacks)
    {
        var existing = (target.StatusEffects ?? new()).ToList();
        var idx = existing.FindIndex(s => s.Id == def.Id);

        if (idx >= 0)
        {
            existing[idx] = existing[idx] with { Stacks = existing[idx].Stacks + stacks };
        }
        else
        {
            existing.Add(new StatusEffect
            {
                Id = def.Id,
                Name = def.Name,
                Icon = def.Icon,
                Stacks = stacks,
                TurnsRemaining = null,
                TriggerMoment = def.TriggerMoment,
                EffectId = def.TickEffectId
            });
        }

        return CardEffectService.CloneCompanion(target, c => c.StatusEffects = existing);
    }

    private static BattleEnemy UpsertStatusEnemy(BattleEnemy target, StatusEffectDefinition def, int stacks)
    {
        var existing = (target.StatusEffects ?? new()).ToList();
        var idx = existing.FindIndex(s => s.Id == def.Id);

        if (idx >= 0)
        {
            existing[idx] = existing[idx] with { Stacks = existing[idx].Stacks + stacks };
        }
        else
        {
            existing.Add(new StatusEffect
            {
                Id = def.Id,
                Name = def.Name,
                Icon = def.Icon,
                Stacks = stacks,
                TurnsRemaining = null,
                TriggerMoment = def.TriggerMoment,
                EffectId = def.TickEffectId
            });
        }

        return CloneBattleEnemy(target, e => e.StatusEffects = existing);
    }

    private static GameState DecayStatus(
        GameState state, string targetId, string statusId, int decay, string kind)
    {
        List<StatusEffect> Updater(List<StatusEffect> statuses) =>
            statuses
                .Select(s =>
                {
                    if (s.Id != statusId) return s;
                    var newStacks = s.Stacks - decay;
                    var newTurns = s.TurnsRemaining.HasValue ? s.TurnsRemaining.Value - 1 : (int?)null;
                    return s with { Stacks = newStacks, TurnsRemaining = newTurns };
                })
                .Where(s => s.Stacks > 0 && (s.TurnsRemaining is null || s.TurnsRemaining > 0))
                .ToList();

        if (kind == "companion")
        {
            return new GameState
            {
                Player = state.Player,
                Graph = state.Graph,
                Cards = state.Cards,
                Companions = state.Companions.Select(c =>
                    c.Id != targetId ? c : CardEffectService.CloneCompanion(c, comp =>
                        comp.StatusEffects = Updater(comp.StatusEffects ?? new()))
                ).ToList(),
                History = state.History,
                Battle = state.Battle,
                PendingAbilityChoices = state.PendingAbilityChoices
            };
        }

        return new GameState
        {
            Player = state.Player,
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = state.Companions,
            History = state.History,
            Battle = state.Battle is not null
                ? CloneBattle(state.Battle, b =>
                    b.Enemies = b.Enemies.Select(e =>
                        e.Id != targetId ? e : CloneBattleEnemy(e, be =>
                            be.StatusEffects = Updater(be.StatusEffects ?? new()))
                    ).ToList())
                : null,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    private static Companion? FindCompanion(GameState state, string id) =>
        state.Companions.FirstOrDefault(c => c.Id == id);

    private static BattleEnemy? FindEnemy(GameState state, string id) =>
        state.Battle?.Enemies.FirstOrDefault(e => e.Id == id);

    internal static BattleState CloneBattle(BattleState src, Action<BattleState>? mutate = null)
    {
        var b = new BattleState
        {
            Active = src.Active,
            Enemies = new List<BattleEnemy>(src.Enemies),
            Turn = src.Turn,
            Log = new List<string>(src.Log),
            PendingCardRewards = new List<PendingCardReward>(src.PendingCardRewards),
            LastTurnActions = src.LastTurnActions is not null ? new List<EnemyTurnAction>(src.LastTurnActions) : null
        };
        mutate?.Invoke(b);
        return b;
    }

    internal static BattleEnemy CloneBattleEnemy(BattleEnemy src, Action<BattleEnemy>? mutate = null)
    {
        var e = new BattleEnemy
        {
            Id = src.Id,
            DefinitionId = src.DefinitionId,
            Name = src.Name,
            Life = src.Life,
            MaxLife = src.MaxLife,
            Shield = src.Shield,
            Energy = src.Energy,
            MaxEnergy = src.MaxEnergy,
            Element = src.Element,
            Type = src.Type,
            Level = src.Level,
            ExpReward = src.ExpReward,
            Rewards = new List<EnemyReward>(src.Rewards),
            AttackSummaries = src.AttackSummaries is not null ? new List<AttackSummary>(src.AttackSummaries) : null,
            KilledByCompanionId = src.KilledByCompanionId,
            StatusEffects = src.StatusEffects is not null ? new List<StatusEffect>(src.StatusEffects) : null
        };
        mutate?.Invoke(e);
        return e;
    }
}
